src/store/RelayDeferredFragmentState.js

```diff
@@ -0,0 +1,130 @@
+/**
+ * Copyright 2013-2015, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule RelayDeferredFragmentState
+ * @flow
+ * @typechecks
+ */
+
+'use strict';
+
+import type {DataID} from 'RelayInternalTypes';
+import type GraphQLDeferredQueryTracker from 'GraphQLDeferredQueryTracker';
+import type RelayPendingQueryTracker from 'RelayPendingQueryTracker';
+
+var forEachObject = require('forEachObject');
+
+export type RelayDeferredFragmentStateOptions = {
+  onSuccess: () => void;
+  onFailure: () => void,
+};
+
+class RelayDeferredFragmentState {
+  _deferredErrors: {[key: string]: Error};
+  _deferredSubscriptions: {[key: string]: any};
+  _deferredQueryTracker: GraphQLDeferredQueryTracker;
+  _options: RelayDeferredFragmentStateOptions;
+  _pendingQueryTracker: RelayPendingQueryTracker;
+
+  constructor(
+    deferredQueryTracker: GraphQLDeferredQueryTracker,
+    pendingQueryTracker: RelayPendingQueryTracker,
+    options: RelayDeferredFragmentStateOptions = {
+      onSuccess: () => {},
+      onFailure: () => {},
+    }
+  ) {
+    this._deferredErrors = {};
+    this._deferredSubscriptions = {};
+    this._deferredQueryTracker = deferredQueryTracker;
+    this._options = options;
+    this._pendingQueryTracker = pendingQueryTracker;
+  }
+
+  getFragmentError(dataID: DataID, fragmentID: string): ?Error {
+    var subscriptionKey = getSubscriptionKey(dataID, fragmentID);
+    return this._deferredErrors[subscriptionKey];
+  }
+
+  hasFragmentData(dataID: DataID, fragmentID: string): boolean {
+    var hasKeys = false;
```

Comment:
I'm not sure how I feel about the Relay pattern of setting objects to nil to make checks faster... but happy to revert that.

(Maybe it would make sense to streamline that pattern with some helpers, at the very least...)